hilti/toolchain/src/compiler/global-optimizer.cc

```diff
@@ -753,6 +753,155 @@ struct ConstantFoldingVisitor : OptimizerVisitor, visitor::PreOrder<bool, Consta
     }
 };
 
+// This visitor collects requirement attributes in the AST and toggles unused features.
+struct FeatureRequirementsVisitor : visitor::PreOrder<void, FeatureRequirementsVisitor> {
+    // Lookup table for feature name -> required.
+    using Features = std::map<std::string, bool>;
+
+    // Lookup table for typename -> features.
+    std::map<ID, Features> _features;
+
+    enum class Stage { COLLECT, TRANSFORM };
+    Stage _stage = Stage::COLLECT;
+
+    void collect(Node& node) {
+        _stage = Stage::COLLECT;
+
+        for ( auto i : this->walk(&node) )
+            dispatch(i);
+    }
+
+    void transform(Node& node) {
+        _stage = Stage::TRANSFORM;
+
+        for ( auto i : this->walk(&node) )
+            dispatch(i);
+    }
+
+    void operator()(const declaration::GlobalVariable& x, position_t p) {
+        const auto& id = x.id();
+
+        // We only work on feature flags named `__feat*`.
+        if ( ! util::startsWith(id, "__feat") )
+            return;
+
+        const auto& tokens = util::split(id, "%");
+        assert(tokens.size() == 3);
+
+        // The type name is encoded in the variable name with `::` replaced by `__`.
+        const auto& typeID = ID(util::replace(tokens[1], "__", "::"));
+        const auto& feature = tokens[2];
+
+        switch ( _stage ) {
+            case Stage::COLLECT: {
+                // Record the feature as unused for the type if it was not already recorded.
+                _features[typeID].insert({feature, false});
+                break;
+            }
+
+            case Stage::TRANSFORM: {
+                const auto required = _features.at(typeID).at(feature);
+                const auto value = x.init().value().as<expression::Ctor>().ctor().as<ctor::Bool>().value();
+
+                if ( required != value ) {
+                    HILTI_DEBUG(logging::debug::GlobalOptimizer,
+                                util::fmt("disabling feature '%s' of type '%s' since it is not used", feature, typeID));
+
+                    auto new_x = declaration::GlobalVariable::setInit(x, builder::bool_(false));
+                    replaceNode(p, new_x);
+                }
+
+                break;
+            }
+        }
+    }
+
+    template<typename MemberAccess>
```

Comment:
Does this need to be templated ? (vs either passing in a `ResolvedOperator` or just the specific pieces directly that are needed).

Reply:
`ResolvedOperator` seems to work fine.